corokafka/impl/corokafka_header_pack_impl.h

```diff
@@ -31,33 +33,51 @@ HeaderPack& HeaderPack::push_back(const std::string& name, H&& header) {
 
 template <typename H>
 const H& HeaderPack::get(const std::string& name, size_t relativePosition) const & {
+    if (!isValid(name, relativePosition)) {
+        throw InvalidArgumentException(0, "Header is invalid (empty)");
+    }
     return boost::any_cast<const H&>(getImpl(name, relativePosition)->second);
 }
 
 template <typename H>
 H& HeaderPack::get(const std::string& name, size_t relativePosition) & {
+    if (!isValid(name, relativePosition)) {
+        throw InvalidArgumentException(0, "Header is invalid (empty)");
+    }
     return boost::any_cast<H&>(getImpl(name, relativePosition)->second);
 }
 
 template <typename H>
 H&& HeaderPack::get(const std::string& name, size_t relativePosition) && {
+    if (!isValid(name, relativePosition)) {
+        throw InvalidArgumentException(0, "Header is invalid (empty)");
+    }
     return boost::any_cast<H&&>(std::move(getImpl(name, relativePosition)->second));
 }
 
 template <typename H>
 HeaderRef<const H&> HeaderPack::getAt(size_t index) const & {
+    if (!isValidAt(index)) {
+        throw InvalidArgumentException(0, "Header is invalid (empty)");
+    }
     const auto& entry = _headers.at(index);
     return HeaderRef<const H&>(entry.first, boost::any_cast<const H&>(entry.second));
 }
 
 template <typename H>
 HeaderRef<H&> HeaderPack::getAt(size_t index) & {
+    if (!isValidAt(index)) {
+        throw InvalidArgumentException(0, "Header is invalid (empty)");
+    }
     auto& entry = _headers.at(index);
     return HeaderRef<H&>(entry.first, boost::any_cast<H&>(entry.second));
 }
 
 template <typename H>
 HeaderRef<H&&> HeaderPack::getAt(size_t index) && {
+    if (!isValidAt(index)) {
+        throw InvalidArgumentException(0, "Header is invalid (empty)");
+    }
```

Comment:
To minimize code repetition,  combine the validation check of isValidAt and isValid to validateHeader functions.  Then only one validate function to write test-cases and future changes to validation logic are in one location.